drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,512 @@
+		return -EINVAL;
+	}
+
+	while (channels) {
+		channel = find_lsb_set(channels) - 1;
+
+		channel_cfg = &data->channels_cfg[channel];
+		channel_cfg->func = func;
+		channel_cfg->param = param;
+
+		WRITE_BIT(channels, channel, 0);
+	}
+
+	return 0;
+}
+
+int adc_emul_ref_voltage_set(const struct device *dev, enum adc_reference ref,
+			     uint16_t value)
+{
+	struct adc_driver_api *api = (struct adc_driver_api *)dev->api;
+	struct adc_emul_data *data = dev->data;
+
+	switch (ref) {
+	case ADC_REF_VDD_1:
+		data->ref_vdd = value;
+		break;
+	case ADC_REF_INTERNAL:
+		data->ref_int = value;
+		api->ref_internal = value;
+		break;
+	case ADC_REF_EXTERNAL0:
+		data->ref_ext0 = value;
+		break;
+	case ADC_REF_EXTERNAL1:
+		data->ref_ext1 = value;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Convert @p ref to reference voltage value in mV
+ *
+ * @param data Internal data of ADC emulator
+ * @param ref Select which reference source should be used
+ *
+ * @return Reference voltage in mV
+ * @return 0 on error
+ */
+static uint16_t adc_emul_get_ref_voltage(struct adc_emul_data *data,
+					 enum adc_reference ref)
+{
+	switch (ref) {
+	case ADC_REF_VDD_1:
+		return data->ref_vdd;
+	case ADC_REF_VDD_1_2:
+		return data->ref_vdd / 2;
+	case ADC_REF_VDD_1_3:
+		return data->ref_vdd / 3;
+	case ADC_REF_VDD_1_4:
+		return data->ref_vdd / 4;
+	case ADC_REF_INTERNAL:
+		return data->ref_int;
+	case ADC_REF_EXTERNAL0:
+		return data->ref_ext0;
+	case ADC_REF_EXTERNAL1:
+		return data->ref_ext1;
+	default:
+		return 0;
+	}
+}
+
+static int adc_emul_channel_setup(const struct device *dev,
+				  const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_channel_cfg *emul_channel_cfg;
+	struct adc_emul_data *data = dev->data;
+
+	if (channel_cfg->channel_id >= config->channels_num) {
+		LOG_ERR("unsupported channel id '%d'", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	emul_channel_cfg = &data->channels_cfg[channel_cfg->channel_id];
+
+	emul_channel_cfg->gain = channel_cfg->gain;
+
+	emul_channel_cfg->ref = channel_cfg->reference;
+	if (adc_emul_get_ref_voltage(data, emul_channel_cfg->ref) == 0) {
+		LOG_ERR("unsupported channel reference '%d'",
+			channel_cfg->reference);
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("unsupported differential mode");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Check if buffer in @p sequence is big enough to hold all ADC samples
+ *
+ * @param dev ADC emulator device
+ * @param sequence ADC sequence description
+ *
+ * @return 0 on success
+ * @return -ENOMEM if buffer is not big enough
+ */
+static int adc_emul_validate_buffer_size(const struct device *dev,
+					 const struct adc_sequence *sequence)
+{
+	const struct adc_emul_config *config = dev->config;
+	uint8_t channels = 0;
+	size_t needed;
+	uint32_t mask;
+
+	for (mask = BIT(config->channels_num - 1); mask != 0; mask >>= 1) {
```

Comment:
Does zephyr have a function to count the set bits? I didn't see it last time I looked though.

Reply:
I didn't found such function either.